exit-from-dialog.js

```diff
@@ -0,0 +1,165 @@
+/*
+  ====================================
+  DIALOG EXIT FUNCTION MOD (@mildmojo)
+  ====================================
+
+  Lets you exit to another room from dialog (including inside conditionals). Use
+  it to make an invisible sprite that acts as a conditional exit, use it to warp
+  somewhere after a conversation, use it to put a guard at your gate who only
+  lets you in once you're disguised, use it to require payment before the
+  ferryman will take you across the river.
+
+  Using the {exitAfter} function in any part of a series of dialog will make the
+  game exit to the new room after the dialog is finished. Using {exitNow} will
+  cut off the dialog text and immediately warp to the new room.
+
+  Usage: {exitAfter "<room name>,<x>,<y>"}
+         {exitAfter "<room name>,<x>,<y>,sprite"}
+         {exitNow "<room name>,<x>,<y>"}
+         {exitNow "<room name>,<x>,<y>,sprite"}
+
+  Example: {exitAfter "FinalRoom,8,4"}
+           {exitNow "FinalRoom,8,11,sprite"}
+
+  WARNING: In exit coordinates, the TOP left tile is (0,0). In sprite coordinates,
+           the BOTTOM left tile is (0,0). If you'd like to use sprite coordinates,
+           add the word "sprite" as the fourth parameter to the exit function.
+
+  Installation:
+    1. Paste all this code into new script tags at the bottom of your exported
+       game's HTML file, right after the last /script> tag.
+
+  NOTE: DON'T EDIT DIALOG FOR SPRITES/ITEMS WITH {exit} CALLS IN THE DIALOG WINDOW.
+        Always edit them in the dialog textbox of the sprite/item window.
+        Editing a sprite or item's dialog in the dialog window will replace your
+        `{exitAfter "room,5,6"}` with `{}` and you probably won't notice that your
+        exits are busted.
+
+        For full editor integration, you'd *probably* also need to paste this
+        code at the end of the editor's `bitsy.js` file. Untested.
+
+  Version: 2.0
+  Bitsy Version: 4.5, 4.6
+  License: WTFPL (do WTF you want) except the `_inject` function by @seleb
+*/
+
+// Give a hoot, don't pollute; encapsulate in an IIFE for isolation.
+(function(globals) {
+  'use strict';
+
+  var queuedDialogExit = null;
+
+  // Hook into the game reset and make sure exit data gets cleared.
+  var _clearGameData = clearGameData;
+  globals.clearGameData = function() {
+    _clearGameData.apply(this, arguments);
+    queuedDialogExit = null;
+  };
+
+  // Hook into the dialog finish event; if there was an {exit}, travel there now.
+  var _onExitDialog = onExitDialog;
+  globals.onExitDialog = function() {
+    _onExitDialog.apply(this, arguments);
+    if (queuedDialogExit) {
+      doPlayerExit(queuedDialogExit);
+      queuedDialogExit = null;
+    }
+  };
+
+  // Implement the {exitAfter} dialog function. It saves the room name and
+  // destination X/Y coordinates so we can travel there after the dialog is over.
+  globals.exitAfterFunc = function(environment, parameters, onReturn) {
```

Comment:
it'd probably be safer to inject the actual function code for the exits into the script instead of putting them in global scope and injecting references to them, e.g.
```js
var exitAfterFunc = function(){...};
...
_inject(
'var functionMap = new Map();',
'functionMap.set("exitAfter", ' + exitAfterFunc.toString() + ');'
);
```

Reply:
Injecting the source of the function inline won't give it access to functions/variables in my enclosure unless I expose them globally. Somethin's gotta be global! 🙂

Reply:
ah good point; you could inject the whole kit and caboodle but for simplicity's sake this makes sense then